pyresample/geometry.py

```diff
@@ -2772,40 +2772,49 @@ def get_geostationary_angle_extent(geos_area):
     aeq = 1 - req ** 2 / (h ** 2)
     ap_ = 1 - rp ** 2 / (h ** 2)
 
-    # generate points around the north hemisphere in satellite projection
-    # make it a bit smaller so that we stay inside the valid area
-    xmax = np.arccos(np.sqrt(aeq))
-    ymax = np.arccos(np.sqrt(ap_))
-    return xmax, ymax
+    x_angle = np.arccos(np.sqrt(aeq))
+    y_angle = np.arccos(np.sqrt(ap_))
+    return x_angle, y_angle
 
 
 def get_geostationary_bounding_box_in_proj_coords(geos_area, nb_points=50):
     """Get the bbox in geos projection coordinates of the valid pixels inside `geos_area`.
 
-    Notes:
-    - The first and last element of the output vectors are equal.
-    - If nb_points is even, it will return x and y vectors of length nb_points + 1.
-
     Parameters
     ----------
     nb_points : Number of points on the polygon.
 
     """
-    xmax, ymax = get_geostationary_angle_extent(geos_area)
+    x, y = get_full_geostationary_bounding_box_in_proj_coords(geos_area, nb_points)
+    ll_x, ll_y, ur_x, ur_y = geos_area.area_extent
+
+    from shapely.geometry import Polygon
+    geo_bbox = Polygon(np.vstack((x, y)).T)
+    area_bbox = Polygon(((ll_x, ll_y), (ll_x, ur_y), (ur_x, ur_y), (ur_x, ll_y)))
+    intersection = area_bbox.intersection(geo_bbox)
```

Comment:
There is a very good chance I'm missing something, but could this be simplified (logically, maybe not in code) to min/max each of the coordinates in the full geostationary bounding box with the extents of the area? So if the full bbox says a point is at `(10_000, -85_000_000)` and the extents of the area are `(-10_000, -60_000_000, 50_000, 10_000_000)` then we know that the full bbox point's `x` coordinate is within the extent, but the `y` coordinate can be clipped to `-60_000_000`.

Not sure this is accurate, or makes sense, or is a good idea, but it is the first thing that comes to mind.

Reply:
ok, so the polygon for the full disk looks like this:
![full_disk](https://user-images.githubusercontent.com/167802/216694816-ff2b1207-f684-4b9b-9d43-8f1eed8e1a9c.png)

Now, if we just clip y at 2e6 like it is in main, we get this:
![clipped_y](https://user-images.githubusercontent.com/167802/216694914-1ff73d65-0a63-49e0-a3f2-518135dbf858.png)

If you look carefully, you see the last point on each side of the y=2e6 line ends up in space, because we 'project' the extreme points of the equator onto a latitude where the sphere is not as wide.

So how do we avoid this? 
One solution would be to not compute everything below y=2e6 and find the edge of the disk at that y position to close the polygon. It's not too hard, but there are many different case, whether the disk is cut horizontally, vertically, or both, includes the equator or prime meridian, or not.
A simpler geometrical approach in my opinion is just to find the intersection of the full disk polygon with the rectangle of the area extent. Since they are all in the same projection, we can see that as a simple 2d geometry problem.
![full_and_box](https://user-images.githubusercontent.com/167802/216699413-0833cade-4032-439a-9531-7a3ffc53e9df.png)

Reply:
Ok, I think I got it. One more alternative solution: mask out (remove) any points not within the bbox. The main point here is that our extent bbox is not a complicated polygon so we only have two limits to check:

```
disk_points = ...
area_points = disk_points[disk_points[:, 0] >= min(extent[0], extent[2]) & disk_points[:, 0] <= max(extent[0], extent[2]) & disk_points[:, 1] >= min(extent[1], extent[3]) & disk_points[:, 1] <= min(extent[1], extent[3])]
```

Reply:
Ugh, nevermind this doesn't get you the very edge of the polygon (the part of the line going to the next outside point). Hm...shoot.

Reply:
Yes, that's why I think we need an intersection to go to the very edge...

Reply:
The only other thing, which is at this point getting ridiculous and we should just use shapely, would be to combine my two suggestions and every outer disk pixel gets clipped to the min/max of the inner disk pixels (the pixels inside the area's bbox).

Reply:
Nevermind again, this last suggestion doesn't fix the issue. You still end up with the points being no closer to the extent of the area. Ok, shapely it is. Maybe at this point we say shapely is required? What is it being used for now? Did @ghiggi want to depend on it more heavily in the future for spherical stuff?

Reply:
Yes, I plan to use it for several tasks. Could make sense to have shapely as a requirement. 
Just check if we need to constrain to shapely < 2.0.0 or we ask for shapely > 2.0.0

Reply:
> Maybe at this point we say shapely is required? What is it being used for now?

Gradient search resampler is using `shapely` to find the overlapping source and target area chunk pairs so that only those are handled.

Reply:
Made shapely a hard requirement